src/utils/stackingUtils.ts

```diff
@@ -0,0 +1,95 @@
+    export module Stacking {
+
+      export type StackedDatum = {
+        value: number;
+        offset: number;
+      };
+
+      export type StackInformation = Utils.Map<Dataset, Utils.Map<string, StackedDatum>>;
+
+      var nativeMath: Math = (<any>window).Math;
+
+      /**
+       * Computes the stacking information (value and offset) for each data point in each dataset
+       */
+      export function computeStackInformation(
+          datasets: Dataset[],
+          keyAccessor: Accessor<any>,
+          valueAccessor: Accessor<number>): StackInformation {
+        var positiveOffsets = d3.map<number>();
+        var negativeOffsets = d3.map<number>();
+        var datasetToKeyToStackedDatum = new Utils.Map<Dataset, Utils.Map<string, StackedDatum>>();
+
+        datasets.forEach((dataset) => {
+          var keyToStackedDatum = new Utils.Map<string, StackedDatum>();
+          dataset.data().forEach((datum, index) => {
+            var key = normalizeKey(keyAccessor(datum, index, dataset));
+            var value = +valueAccessor(datum, index, dataset);
+            var offset: number;
+            var offsetMap = (value >= 0) ? positiveOffsets : negativeOffsets;
+            if (offsetMap.has(key)) {
+              offset = offsetMap.get(key);
+              offsetMap.set(key, offset + value);
+            } else {
+              offset = 0;
+              offsetMap.set(key, value);
+            }
+            keyToStackedDatum.set(key, {
+              value: value,
+              offset: offset
+            });
+          });
+          datasetToKeyToStackedDatum.set(dataset, keyToStackedDatum);
+        });
+        return datasetToKeyToStackedDatum;
+      }
+
+      /**
+       * Computes the total extent over all data points in all datasets, taking stacking into consideration
+       */
+      export function computeStackExtent(stackInformation: StackInformation, keyAccessor: Accessor<any>, filter: Accessor<boolean>) {
+        var extents: number[] = [];
+        stackInformation.forEach((stackedDatumMap: Utils.Map<string, StackedDatum>, dataset: Dataset) => {
+          dataset.data().forEach((datum, index) => {
+            if (filter != null && !filter(datum, index, dataset)) {
+              return;
+            }
+            var stackedDatum = stackedDatumMap.get(normalizeKey(keyAccessor(datum, index, dataset)));
+            extents.push(stackedDatum.value + stackedDatum.offset);
+          });
+        });
+        var maxStackExtent = Utils.Math.max(extents, 0);
+        var minStackExtent = Utils.Math.min(extents, 0);
+
+        return [nativeMath.min(minStackExtent, 0), nativeMath.max(0, maxStackExtent)];
+      }
+
+      /**
+       * Given an array of datasets and the accessor function for the key, computes the
+       * set reunion (no duplicates) of the domain of each dataset.
+       */
+      export function domainKeys(datasets: Dataset[], keyAccessor: Accessor<any>) {
+        var domainKeys = d3.set();
```

Comment:
I'll note that this method normalizes the keys by stringifying them, but there's no warning of this for the user.